### PR TITLE
xwayland-satellite: init at 0.4

### PR DESCRIPTION
**Pull request by if-loop69420**
## Description of changes

Added xwayland-satellite, which grants rootless Xwayland integration to any Wayland compositor implementing xdg_wm_base. This is particularly useful for compositors that (understandably) do not want to go through implementing support for rootless Xwayland themselves.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/xw/xwayland-satellite/package.nix [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  ];
+
+  # Highly impure tests, which fail to run in the sandbox
+  doCheck = false;
+  # checkFlags = [
+  #   "--skip=xwayland-satellite::tests::integration"
+  # ];
+
+  meta = with lib; {
+    description = "Program for rootless Xwayland integration to any Wayland compositor";
```
Comment (christoph-heiss) [inline pkgs/by-name/xw/xwayland-satellite/package.nix:52]:
```suggestion
    description = "Rootless Xwayland integration to any Wayland compositor implementing xdg_wm_base";
```

File: pkgs/by-name/xw/xwayland-satellite/package.nix [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    xwayland
+    libxcb
+    xcb-util-cursor
+  ];
+
+  # Highly impure tests, which fail to run in the sandbox
+  doCheck = false;
+  # checkFlags = [
+  #   "--skip=xwayland-satellite::tests::integration"
+  # ];
```
Comment (christoph-heiss) [inline pkgs/by-name/xw/xwayland-satellite/package.nix:49]: Only the actual integration tests need to be disabled:
```suggestion
  # disable Xwayland integration tests which need a running display server 
  checkFlags = [
    "--exact"
    "--skip=copy_from_wayland"
    "--skip=copy_from_x11"
    "--skip=input_focus"
    "--skip=quick_delete"
    "--skip=reparent"
    "--skip=toplevel_flow"
  ];
```

**Review by christoph-heiss — CHANGES_REQUESTED** [review]
Please consolidate the commits into just two:
- adding the maintainer entry (w/ a commit message like `maintainers: add <username>`)
- package init (commit message like the PR title)

Result of `nixpkgs-review pr 322396` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>xwayland-satellite</li>
  </ul>
</details>

Binary LGTM, quickly tested it out.

File: pkgs/by-name/xw/xwayland-satellite/package.nix [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  src = fetchFromGitHub {
+    owner = "Supreeeme";
+    repo = "xwayland-satellite";
+    rev = "v${version}";
+    hash = "sha256-dwF9nI54a6Fo9XU5s4qmvMXSgCid3YQVGxch00qEMvI=";
+  };
+
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+  };
```
Comment (christoph-heiss) [inline pkgs/by-name/xw/xwayland-satellite/package.nix:26]: Can be simplified as xwayland-satellite does not use git dependencies:
```suggestion
  cargoHash = "<whatever>";
```

**Review by christoph-heiss — CHANGES_REQUESTED** [review]
Also nix-update can be enabled using:
```nix
passthru.updateScript = nix-update-script { };
```

Normally, this is added before the `meta` attribute.

**Comment by if-loop69420** [comment]
Fixed all suggestions you had

**Review by christoph-heiss — APPROVED** [review]

File: pkgs/by-name/xw/xwayland-satellite/package.nix [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    hash = "sha256-dwF9nI54a6Fo9XU5s4qmvMXSgCid3YQVGxch00qEMvI=";
+  };
+
+  cargoHash = "sha256-nKPSkHbh73xKWNpN/OpDmLnVmA3uygs3a+ejOhwU3yA=";
+
+  LIBCLANG_PATH = "${llvmPackages.libclang.lib}/lib";
+
+  nativeBuildInputs = [
+    pkg-config
+    clang
```
Comment (Aleksanaa) [inline pkgs/by-name/xw/xwayland-satellite/package.nix:31]: check `rustPlatform.bindgenHook`. If that works remove all clang usage.

**Comment by if-loop69420** [comment]
Done

File: pkgs/by-name/xw/xwayland-satellite/package.nix [hunk 5]
```diff
@@ -0,0 +1,59 @@
+{ lib
+, rustPlatform
+, fetchFromGitHub
+, pkg-config
+, clang
```
Comment (christoph-heiss) [inline pkgs/by-name/xw/xwayland-satellite/package.nix:5]: Unused, please drop.

File: pkgs/by-name/xw/xwayland-satellite/package.nix [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+, xwayland
+, xorg
```
Comment (christoph-heiss) [inline pkgs/by-name/xw/xwayland-satellite/package.nix:7]: Unused, please drop.

File: pkgs/by-name/xw/xwayland-satellite/package.nix [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+, xcb-util-cursor
+, libxcb
+, llvmPackages
```
Comment (christoph-heiss) [inline pkgs/by-name/xw/xwayland-satellite/package.nix:10]: Unused, please drop.

**Review by christoph-heiss — CHANGES_REQUESTED** [review]

**Comment by Aleksanaa** [comment]
Please reword it to "xwayland-satellite: init at 0.4" otherwise ofBorg won't build

**Comment by if-loop69420** [comment]
What do you mean by reword "it"?

**Comment by Aleksanaa** [comment]
The commit https://github.com/NixOS/nixpkgs/pull/322396/commits/9619ad71bd24811f47371167c513ed725df098a7

**Comment by if-loop69420** [comment]
Done

**Comment by Aleksanaa** [comment]
Ran it but don't know how to test it. Could you explain that?

**Comment by if-loop69420** [comment]
The very bottom of this page explains it https://github.com/YaLTeR/niri/wiki/Xwayland. 

Basically launch xwayland-satellite and then run any program, that uses x11 like this: ```env DISPLAY=:0 <command>```.

Note: This project is very early into it's development and of course has some rough edges. Steam for example doesn't run very stable through xwayland-satellite on my machine.

**Comment by Aleksanaa** [comment]
The problem is I'm using hyprland at the moment and it doesn't seem to use this program to render xwayland. If you've tested it I'll merge it.

**Comment by if-loop69420** [comment]
If you need a concrete example install GeoGebra through flatpak (```flatpak install org.geogebra.GeoGebra```) and then run it using ```env DISPLAY=:0 flatpak run org.geogebra.GeoGebra```. It should then run in an xwayland window through xwayland-satellite. I've tested this on niri, which doesn't have a built in mechanism for running things in xwayland. I have tested it with steam and geogebra and it works

**Review by Aleksanaa — APPROVED** [review]

**Comment by if-loop69420** [comment]
I can add a screenshot of GeoGebra running this way if needed

**Comment by if-loop69420** [comment]
Ah perfect. Thanks :)

**Comment by if-loop69420** [comment]
@midirhee12 I don't know, what i am doing differently to your setup. My config is available in if-loop69420/cornflake if you want to have a look. I think there was documentation from niri or xwayland-satellite itself on how to get it to work.

**Comment by if-loop69420** [comment]
@midirhee12 May also be related to #352725
